### PR TITLE
[Merged by Bors] - feat:  Prove isomorphism implies equality of normal closure subgroups

### PR DESCRIPTION
**Pull request by newell**
feat: Prove isomorphism implies equality of normal closure subgroups

Co-authored-by: Riccardo Brasca riccardo.brasca@gmail.com

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:1506]:
```suggestion
  map_equiv_eq_comap_symm
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3507]: Can you also show the similar (and perhaps easier to prove)
```lean
    (Subgroup.normalClosure s).comap f = Subgroup.normalClosure (f ⁻¹' s) := by
```

Comment (alreadydone) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3507]: Good idea. I think `Subgroup.normalClosure (f ⁻¹' s) ≤ (Subgroup.normalClosure s).comap f` holds for any homomorphism `f : G →* N`.

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (riccardobrasca) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:1504]: In my opinion this should be the umprime version, and we can even drop the other one (I didn't check all the version, but it is coherent with the version for ring equiv).

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:1504]: Yeah, I agree with that based off of what you said (I haven't researched whether this is the general convention) but curious if others agree?  @eric-wieser @alreadydone ?

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:1504]: Will this mess up the #align for mathport (I am completely ignorant of this and whether we can change the names that are currently there -- this would be needed afaict).

Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:1504]: You should adjust the `#align` to point to the best match for the mathlib3 statement, if one exists. So here, either change the align to point to the primed name (if you keep both), or just leave it alone (if you change the existing existing statement as Riccardo suggests)

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3531]:
```suggestion
    (normalClosure s).comap f = normalClosure (f ⁻¹' s) := by
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3512]: I think this would make more sense as
```suggestion
theorem Subgroup.map_normalClosure {s : Set G} (f : G ≃* N) :
    (normalClosure s).map f = normalClosure (f '' s) := by
```
as this matches `map_closure`

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3513]: I pushed a golf here: the result is that this lemma should only need `f` to be surjective, not an equivalence. Can you update things accordingly?

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3513]: I think I originally misunderstood what you meant here and pushed still using `f` as an equivalence.  Will take another stab at it.

Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3513]: What you pushed is what *I* originally meant, but @alreadydone suggested it might be true more generally

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3513]: I was able to prove the lemma that @alreadydone mentioned but not really sure it fits well into the other proofs.  I also generalized to use only the surjection for what you showed above but haven't found out why I can get `map_normalClosure` to work but not `map_normalClosure'`.  At any rate, just let me know what direction you think is best for these lemmas and I will try wrapping it all up in that direction:

```
theorem Subgroup.map_normalClosure {s : Set G} (f : G ≃* N) :
    (normalClosure s).map f = normalClosure (f '' s) := by
  have : Normal (map f (normalClosure s)) := Normal.map inferInstance f f.surjective
  apply le_antisymm
  · rw [map_le_iff_le_comap]
    refine normalClosure_le_normal ?_
    simp only [coe_comap, ← Set.image_subset_iff]
    exact subset_normalClosure
  · exact normalClosure_le_normal (Set.image_subset f subset_normalClosure)

theorem Subgroup.map_normalClosure' {s : Set G} (f : G →* N) (h : Function.Surjective f) :
    (normalClosure s).map f = normalClosure (f '' s) := by
  have : Normal (map f (normalClosure s)) := Normal.map inferInstance f h
  apply le_antisymm
  · rw [map_le_iff_le_comap]
    refine normalClosure_le_normal ?_
    simp only [coe_comap, ← Set.image_subset_iff]
    exact subset_normalClosure
  · exact normalClosure_le_normal (Set.image_subset f subset_normalClosure)

theorem Subgroup.normalClosure_preimage_le_comap {s : Set N} (f : G →* N) :
    normalClosure (f ⁻¹' s) ≤ (normalClosure s).comap f := by
  refine normalClosure_le_normal ?h
  simp only [coe_comap]
  refine Set.preimage_mono ?h.h
  exact subset_normalClosure

theorem Subgroup.comap_normalClosure {s : Set N} (f : G ≃* N) :
    normalClosure (f ⁻¹' s) = (normalClosure s).comap f  := by
  have := Set.preimage_equiv_eq_image_symm s f.toEquiv
  simp_all [comap_equiv_eq_map_symm, map_normalClosure]

theorem Subgroup.comap_normalClosure' {s : Set N} (f : G ≃* N) :
    normalClosure (f ⁻¹' s) = (normalClosure s).comap f  := by
  have := Set.preimage_equiv_eq_image_symm s f.toEquiv
  simp_all [comap_equiv_eq_map_symm]
  -- simp [@map_normalClosure' G _ N _ (f ⁻¹' s) f.toMonoidHom f.surjective]
```

Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3513]: I think my comment above still applies; instead of using `f.surjective`, can you add `hf : Surjective f`, and replace `f` with a `MonoidHom`?

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean]: Yes, update uses a `MonoidHom` and its surjection explicitly now.

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3522]: Is the `@` really needed here?

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (eric-wieser) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3511]:
```suggestion
theorem Subgroup.map_normalClosure (s : Set G) (f : G →* N) (hf : Surjective f) :
```
and below

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3522]: > Is the `@` really needed here?

Nope, removed.

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (tb65536) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:3520]: Maybe this is best saved for another PR, but you can weaken this to `Surjective f` under the assumption that `s` is `Nonempty`:
```
theorem Subgroup.ker_le_closure_preimage_of_nonempty (s : Set N)
    (f : G →* N) (hs : (f ⁻¹' s).Nonempty) : f.ker ≤ closure (f ⁻¹' s) := by
  obtain ⟨g, hg⟩ := hs
  refine' le_trans (fun k h ↦ _) (closure_mono (Set.preimage_mono (Set.singleton_subset_iff.2 hg)))
  rw [← mul_inv_cancel_right k g]
  refine' mul_mem (subset_closure _) (inv_mem (subset_closure _)) <;> simp [map_mul, f.mem_ker.1 h]

theorem Subgroup.comap_normalClosure_of_nonempty (s : Set N) (hs : s.Nonempty)
    (f : G →* N) (hf : Surjective f) :
    normalClosure (f ⁻¹' s) = (normalClosure s).comap f := by
  refine' map_injective_of_ker_le f
    ((ker_le_closure_preimage_of_nonempty s f (hs.preimage hf)).trans closure_le_normalClosure)
    (ker_le_comap f (normalClosure s)) _
  rw [map_normalClosure (f ⁻¹' s) f hf, map_comap_eq_self_of_surjective hf,
    Set.image_preimage_eq s hf]
```

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:1498]: @tb65536 Yeah, @riccardobrasca had a similar sentiment.  I don't have any particularly strong opinions as I am still somewhat new to Mathlib and still learning my way around and open to whatever the reviewers think is best.  @eric-wieser any opinions on this?

File: Mathlib/GroupTheory/Subgroup/Basic.lean
Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3520]: @tb65536 I had looked for a proof to weaken the equivalence assumption so I am glad to see this!  

To further the sake of symmetry among the lemmas it may be possible to remove the `NonEmpty` from the assumptions and then do a proof `by_cases H : s.NonEmpty` *if* when `s` is empty you can derive a contradiction.  It is getting in the wee hours of the morning here so I may take a stab at it in the morning but feel free to add this if you think this is a decent idea (if not, I would like to hear back on the reasons so I can learn why) and go ahead and push a commit.  

Thanks for the feedback!

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3520]: @eric-wieser what should be done here?  I want to get this PR across the line as I have a couple of branches blocked on it.

Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3520]: Given this is not a strict generalization, these should be in addition to the existing lemmas not instead of them. I don't care whether they are in this PR or a follow-up.

Comment (newell) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:3520]: I say we just keep them in a separate PR then.

**Comment by tb65536** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

(Future TODO: someone should clean up the variable names. `G` and `N` are playing mostly symmetric roles, but their names don't suggest that.)

**Pull request closed without merging**
